protostar/io/output.py

```diff
@@ -39,7 +62,28 @@ def __call__(self, message: Message):
                 f"JSON output is not supported for messages of type {type(message).__name__}."
             )
 
-        dictionary = message.format_dict()
-        json.dump(dictionary, sys.stdout, allow_nan=False, separators=(",", ":"))
+        obj = message.format_dict()
+        self._print(obj)
+
+    @contextmanager
+    def activity(self, message_template: Message) -> Generator[None, None, None]:
+        if not isinstance(message_template, StructuredMessage):
+            raise NotImplementedError(
+                f"JSON output is not supported for messages of type {type(message_template).__name__}."
+            )
+
+        rendered = message_template.format_dict()
+
+        self._print({**rendered, "event": "started"})
+        try:
+            yield
+            self._print({**rendered, "event": "ok"})
+        except BaseException:
+            self._print({**rendered, "event": "failed"})
+            raise
```

Comment:
(We need to figure out how to display errors (ProtostarException), but this is out of scope of this PR.)

Reply:
#1155